### ISSUE TITLE
S3 fixture in test_storage.py may not be reliable

### ISSUE THREAD

**Issue by andrewfulton9** [zarr-developers/zarr-python#652]
`test_storage.py:TestFSStore:test_s3_complex` is having unexpected failures when working on zarr. I first had issues with this while working on [PR#645](https://github.com/zarr-developers/zarr-python/pull/645). In that case, when [`a` in test_s3_complex was trying to be reset](https://github.com/zarr-developers/zarr-python/blob/847e78a26270dbb2114771d820977eb5c2d3c5d4/zarr/tests/test_storage.py#L997), it failed to do so unless called twice. This wouldn't happen if I removed the read only group `g2`. I ended up fixing this issue in my last [PR by adding the logic to keep trying to delete the dataset until it didn't exist anymore](https://github.com/zarr-developers/zarr-python/blob/847e78a26270dbb2114771d820977eb5c2d3c5d4/zarr/storage.py#L1117). 

Now I am working in a new branch, and am again getting an error with this test. This time, it is failing to set [`a[:] = 5`](https://github.com/zarr-developers/zarr-python/blob/847e78a26270dbb2114771d820977eb5c2d3c5d4/zarr/tests/test_storage.py#L982). Again, this doesn't happen if I remove the read only `g2` group logic. I spent quite a bit of time debugging, and as far as I can tell, all the keys are passed to fsspec properly to be set to 5, but [only the values of `a` that are already explicitly set](https://github.com/zarr-developers/zarr-python/blob/847e78a26270dbb2114771d820977eb5c2d3c5d4/zarr/tests/test_storage.py#L967-L971), get the value of 5 propagated. 

These errors made me suspect that the s3 fixture isn't working properly, so I ran the test using a real AWS S3 bucket, and in that case the test passes and everything is set properly.

 The odd thing is the fixture throws errors predictably. So while it is failing in a branch, it fails the same way every time, and in my experience with PR#645, regardless of the computer/environment (same failure on my computer and in jenkins). Adding `time.sleep(n)` between setting values and assertions doesn't resolve the issue for me. I also have tried deleting and reinstalling my environment, and the failures still occur as well.

The current branch I am having failures can be found here:
[andrewfulton9/zarr-python branch:partial_read](https://github.com/andrewfulton9/zarr-python/tree/partial_read).

in short: `test_storage.py::TestFSStore::test_s3_complex` is consistently failing, but same code passes when fixture is replaced with real S3 bucket.

#### Version and installation information

Please provide the following:

* Value of ``zarr.__version__``:  branch off current master
* Value of ``numcodecs.__version__``: 0.7.2 (also occured with 0.6.4)
* Version of Python interpreter: >=python3.7
* Operating system (Linux/Windows/Mac): ubuntu 18.04 (also occured on jenkins CI/CD)
* How Zarr was installed (e.g., "using pip into virtual environment", or "using conda")
pip into virtual environment

**Comment by andrewfulton9** [comment]
@martindurant do you have any thoughts on this? I suspect you were right in [your comment on PR 645](https://github.com/zarr-developers/zarr-python/pull/645#discussion_r519025873) and that its related to this.

**Comment by martindurant** [comment]
Throwing an `s3.invalidate_cache` can always be helpful, or turning of s3 directory caching altogether, but I can't really explain the difference in behaviour - especially since this apparently was working recently. Maybe travis' workers have become even slower?

**Comment by andrewfulton9** [comment]
It looks like caching is the issue. If I add the `s3.invalidate`_cache to `test_s3_complex`, it no longer fails. How can I turn off directory caching altogether? I have: 

``` python
s3 = s3fs.S3FileSystem(default_cache_type=None,
                                       default_fill_cache=False,
                                       skip_instance_cache=True,
                                       use_listings_cache=False,
                                       anon=False, **s3so)
```
But it still fails unless I call `invalidate_cache` so I think I must be missing something.

**Comment by andrewfulton9** [comment]
Never mind, I figured it out. That wasn't the right place to declare those to get passed to the `test_s3_complex` groups.

**Comment by andrewfulton9** [comment]
This should be resolved now so closing. Thank you @martindurant for pointing me in the right direction!
